### ISSUE TITLE
PCoA explained variance...

### ISSUE THREAD

**Issue by agartland** [biocore/scikit-bio#1480]
I was taking a look through the source to reconcile some discrepancies in how I was computing explained variance in my own code for PCoA. I noticed that around line 99 of `_principle_coordinate_analysis.py` there is a section on setting negative eigenvalues to zero:

```
# eigvals might not be ordered, so we order them (at least one
# is zero). cogent makes eigenvalues positive by taking the
# abs value, but that doesn't seem to be an approach accepted
# by L&L to deal with negative eigenvalues. We raise a warning
# in that case. First, we make values close to 0 equal to 0.
negative_close_to_zero = np.isclose(eigvals, 0)
eigvals[negative_close_to_zero] = 0
```
The comment suggests we set all values to zero if they are close to zero. The code does this as well, though the variable name suggests only negative values close to zero should be set. Which is correct? Also, it looks like larger negative values might not be changed at all and I think this may impact the explained variance reported later, right? I don't know what is the correct thing to do here, but I'd be interested to hear others' thoughts. Also any intuition about what aspect of the dataset leads to negative eigenvalues would be helpful as well. Thanks!

**Comment by jairideout** [comment]
cc: @Jorge-C, @antgonza, @ElDeveloper

**Comment by antgonza** [comment]
A quick google search yielded these links: 
- https://www.quora.com/When-will-a-matrix-have-negative-eigenvalues-And-what-does-that-mean, does that answer your question? Second answer seems to explain it better.  
- http://stats.stackexchange.com/questions/68680/how-to-interpret-variation-explained-by-principal-coordinates

**Comment by Jorge-C** [comment]
Hi @agartland you're right that the variable name would be more accurate as something like `close_to_zero`.

It doesn't seem unreasonable to me to zero all the negative values when computing the explained variance, but I'm not aware of citations one way or another or if there's a consensus. At any rate, I'm not sure it matters too much because if the negative values are large in absolute terms compared to the positive ones, the pcoa results aren't quite valid, and if they're small then no matter if zeroed or not, the net effect is going to be small.

**Comment by agartland** [comment]
I agree that the impact of this change may be quite small, but now that we understand a bit of what's going on with negative eigenvalues I don't understand why the code would zero out positive eigenvalues that are close to zero (as it does). It would seem to me that small positive values are expected and should be counted as part of the total variance. Also, the `np.isclose` uses an arbitrary relative and absolute tolerance to find values close to zero. If it were up to me I'd set all negative eigenvalues to zero and warn about large negative ones (maybe use a tolerance of 0.5% of total variance?)

**Comment by rob-knight** [comment]
We should clearly benchmark the different approaches on range of datasets and generate such a paper for citation yes?

On Jan 27, 2017, at 10:02 AM, Jorge Ca?ardo Alastuey <notifications@github.com<mailto:notifications@github.com>> wrote:


Hi @agartland<https://github.com/agartland> you're right that the variable name would be more accurate as something like close_to_zero.

It doesn't seem unreasonable to me to zero all the negative values when computing the explained variance, but I'm not aware of citations one way or another or if there's a consensus. At any rate, I'm not sure it matters too much because if the negative values are large in absolute terms compared to the positive ones, the pcoa results aren't quite valid, and if they're small then no matter if zeroed or not, the net effect is going to be small.

-

**Comment by agartland** [comment]
And perhaps such a paper could explore metrics and datasets that lead to negative eigenvalues or pathological distance matrices.
This work may already exist, as I haven't looked extensively yet. I can ask a few colleagues around here to see if they know of any such studies.
